### PR TITLE
Accounting Model Setup

### PR DESCRIPTION
**Pull request by Skyquek**
## Description & Motivation

Added accounting setup django models based on the ERD diagram designed. 

## Related Issue(s)

Nope. 

## Testing

```cmd
python manage.py sqlmigrate accountings 0001 > sql_command_test.sql
```

Basically I dump the SQL generated to the `sql_command_test.sql` to check on the sql generated. 

## Backwards-compatibility

> Is this a breaking change that will not be backwards-compatible? If yes, how so?

Nope

## Documentation

> Does the change require any updates to documentation? If so, where? Are they included?

Nope.

### CODE REVIEW

File: mini-project/BillManagementSystem/accountings/models.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+from djmoney.models.fields import MoneyField
+from phonenumber_field.modelfields import PhoneNumberField
+
+from django.db import models
+from django.contrib.auth.models import User
+
+
+# Create your models here.
+class Payment(models.Model):
+    user_id = models.ForeignKey(User, on_delete=models.CASCADE)
```
Comment (kwongtn) [inline mini-project/BillManagementSystem/accountings/models.py]: With django, we just do like the following:
```py
user = models.ForeignKey(User, on_delete=models.CASCADE)
```

In this case, `user_id` is generally used for internal representation of the fk id itself, as `user` will be used as object representation.

Refer the following on the return objects when called (when using django practices):
```py
pymt_obj = Payment.objects.all()[0]

pymt_obj.user_id   # This returns a str or int
pymt_obj.user      # This returns an instance of the object
```

Same with other instances of this line throughout this model file.

File: mini-project/BillManagementSystem/accountings/models.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+class Category(models.Model):
+    name = models.TextField()
+    
+class Expenses(models.Model):
+    category = models.ForeignKey(Category, on_delete=models.CASCADE)
+    user_id = models.ForeignKey(User, on_delete=models.CASCADE)
+    amount = MoneyField(max_digits=19, decimal_places=2, default_currency='MYR')
+    note = models.TextField()
+    created_at = models.DateTimeField()
+    updated_at = models.DateTimeField()
```
Comment (kwongtn) [inline mini-project/BillManagementSystem/accountings/models.py]: Consider using `TimeStampedModel` from `django_model_utils` to make use of automatic triggers

File: mini-project/BillManagementSystem/sql_command_test.sql [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+--
+CREATE TABLE "accountings_expenses" ("id" integer NOT NULL PRIMARY KEY AUTOINCREMENT, "amount_currency" varchar(3) NOT NULL, "amount" decimal NOT NULL, "note" text NOT NULL, "created_at" datetime NOT NULL, "updated_at" datetime NOT NULL, "category_id" bigint NOT NULL REFERENCES "accountings_category" ("id") DEFERRABLE INITIALLY DEFERRED, "user_id_id" integer NOT NULL REFERENCES "auth_user" ("id") DEFERRABLE INITIALLY DEFERRED);
+--
+-- Create model Debt
+--
+CREATE TABLE "accountings_debt" ("id" integer NOT NULL PRIMARY KEY AUTOINCREMENT, "amount_currency" varchar(3) NOT NULL, "amount" decimal NOT NULL, "created_at" datetime NOT NULL, "updated_at" datetime NOT NULL, "user_id_id" integer NOT NULL REFERENCES "auth_user" ("id") DEFERRABLE INITIALLY DEFERRED);
+CREATE INDEX "accountings_payment_user_id_id_8398c27b" ON "accountings_payment" ("user_id_id");
+CREATE INDEX "accountings_expenses_category_id_54df5b5f" ON "accountings_expenses" ("category_id");
+CREATE INDEX "accountings_expenses_user_id_id_1fa9b981" ON "accountings_expenses" ("user_id_id");
+CREATE INDEX "accountings_debt_user_id_id_2edd167f" ON "accountings_debt" ("user_id_id");
```
Comment (kwongtn) [inline mini-project/BillManagementSystem/sql_command_test.sql]: From here you can see that if you define `user_id` as foreign object, `user_id_id` will be used as the actual foreign key.

File: mini-project/BillManagementSystem/accountings/models.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    
+class Debt(models.Model):
+    user_id = models.ForeignKey(User, on_delete=models.CASCADE)
+    amount = MoneyField(max_digits=19, decimal_places=2, default_currency='MYR')
+    created_at = models.DateTimeField()
+    updated_at = models.DateTimeField()
+    
+class Category(models.Model):
+    name = models.TextField()
```
Comment (kwongtn) [inline mini-project/BillManagementSystem/accountings/models.py]: No unique constraints or of the sort? 😉

File: mini-project/BillManagementSystem/accountings/models.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    category = models.ForeignKey(Category, on_delete=models.CASCADE)
+    user_id = models.ForeignKey(User, on_delete=models.CASCADE)
+    amount = MoneyField(max_digits=19, decimal_places=2, default_currency='MYR')
+    note = models.TextField()
+    created_at = models.DateTimeField()
+    updated_at = models.DateTimeField()
+
+class Student(models.Model):
+    user = models.OneToOneField(User, on_delete=models.CASCADE)
+    birthday = models.DateTimeField()
```
Comment (kwongtn) [inline mini-project/BillManagementSystem/accountings/models.py]: Birthday... `DateTime` field? `Date` should be sufficient

File: mini-project/BillManagementSystem/accountings/models.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    email = models.EmailField()
```
Comment (kwongtn) [inline mini-project/BillManagementSystem/accountings/models.py]: Duplicate of field in django user model

**Review by kwongtn — CHANGES_REQUESTED** [review]
Few things about the entire PR:
- The sqlite3 file is usually not committed.
- We usually use singular for app name -- `accounting` and not `accountings`. Plural naming is used for datatypes like lists, sets etc.

**Comment by Skyquek** [comment]
> * We usually use singular for app name -- `accounting` and not `accountings`. Plural naming is used for datatypes like lists, sets etc.

I see. But to rename the app, all the things need to redo right? Should I just use plural for this time around? Or should I just rename it? 

https://stackoverflow.com/questions/8408046/how-to-change-the-name-of-a-django-app

Its look a bit tedious to rename it.

File: mini-project/BillManagementSystem/accountings/models.py (hunk 1 above)
Comment (Skyquek) [reply mini-project/BillManagementSystem/accountings/models.py]: Noted. Just realized the sql generated will output `user_id_id`.

File: mini-project/BillManagementSystem/sql_command_test.sql (hunk 3 above)
Comment (Skyquek) [reply mini-project/BillManagementSystem/sql_command_test.sql]: Ya, just realize about this.

File: mini-project/BillManagementSystem/accountings/models.py (hunk 5 above)
Comment (Skyquek) [reply mini-project/BillManagementSystem/accountings/models.py]: Ya, silly me. Let me change that! 😳

File: mini-project/BillManagementSystem/accountings/models.py (hunk 6 above)
Comment (Skyquek) [reply mini-project/BillManagementSystem/accountings/models.py]: @kwongtn how can I know the default field available in `User`? Any docs on that?

File: mini-project/BillManagementSystem/accountings/models.py (hunk 4 above)
Comment (Skyquek) [reply mini-project/BillManagementSystem/accountings/models.py]: I added max_length. Is that okay? 😀

**Comment by kwongtn** [comment]
> I see. But to rename the app, all the things need to redo right? Should I just use plural for this time around? Or should I just rename it?
> 
> https://stackoverflow.com/questions/8408046/how-to-change-the-name-of-a-django-app
> 
> Its look a bit tedious to rename it.

I would recommend to rename it. Since migrations are not done (you can always revert it if its done), its relatively simple to rename.
Can also make this a chance to know which is declared where.

Alternatively, you can create a new (correct) app and copy everything there.

File: mini-project/BillManagementSystem/accountings/models.py (hunk 6 above)
Comment (Skyquek) [reply mini-project/BillManagementSystem/accountings/models.py]: Found it . Line 358.
```python
email = models.EmailField(_("email address"), blank=True)
```

File: mini-project/BillManagementSystem/accountings/models.py (hunk 4 above)
Comment (kwongtn) [reply mini-project/BillManagementSystem/accountings/models.py]: No problems with char length actually.
Just that imagine your users may potentially create 100 categories with name "ABC" 🤣

**Comment by Skyquek** [comment]
> I would recommend to rename it. Since migrations are not done (you can always revert it if its done), its relatively simple to rename.

Ya, you are right. It's quite simple to rename it.

File: mini-project/BillManagementSystem/accountings/models.py (hunk 4 above)
Comment (Skyquek) [reply mini-project/BillManagementSystem/accountings/models.py]: Darn, I added a `unique=True` flag then, hope that it works. I can see this line of sql being generated now. 

```sql
CREATE TABLE "new__accounting_category" ("id" integer NOT NULL PRIMARY KEY AUTOINCREMENT, "name" text NOT NULL UNIQUE);
```

@kwongtn What do you think?

**Review by kwongtn — APPROVED** [review]

File: mini-project/BillManagementSystem/accountings/models.py (hunk 4 above)
Comment (kwongtn) [reply mini-project/BillManagementSystem/accountings/models.py]: Looks good

**Pull request merged**